<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.2.6" />
<style type="text/css">
/* Debug borders */
p, li, dt, dd, div, pre, h1, h2, h3, h4, h5, h6 {
/*
  border: 1px solid red;
*/
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

tt {
  color: navy;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  font-family: sans-serif;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}

div.sectionbody {
  font-family: serif;
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}

pre {
  padding: 0;
  margin: 0;
}

span#author {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  font-size: 1.1em;
}
span#email {
}
span#revision {
  font-family: sans-serif;
}

div#footer {
  font-family: sans-serif;
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
div#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
div#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

div#preamble,
div.tableblock, div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-right: 10%;
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.5em;
  margin-bottom: 2.5em;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}

div.listingblock {
  margin-right: 0%;
}
div.listingblock > div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock > div.content {
  padding-left: 2.0em;
}

div.attribution {
  text-align: right;
}
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 2px solid silver;
}

div.exampleblock > div.content {
  border-left: 2px solid silver;
  padding: 0.5em;
}

div.verseblock div.content {
  white-space: pre;
}

div.imageblock div.content { padding-left: 0; }
div.imageblock img { border: 1px solid silver; }
span.image img { border-style: none; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
div.olist > ol {
  list-style-type: decimal;
}
div.olist2 > ol {
  list-style-type: lower-alpha;
}

div.tableblock > table {
  border: 3px solid #527bbd;
}
thead {
  font-family: sans-serif;
  font-weight: bold;
}
tfoot {
  font-weight: bold;
}

div.hlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hlist td {
  padding-bottom: 15px;
}
td.hlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
}
td.hlist2 {
  vertical-align: top;
}

@media print {
  div#footer-badges { display: none; }
}

div#toctitle {
  color: #527bbd;
  font-family: sans-serif;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}
/* Workarounds for IE6's broken and incomplete CSS2. */

div.sidebar-content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}
div.sidebar-title, div.image-title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  margin-top: 0.0em;
  margin-bottom: 0.5em;
}

div.listingblock div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock-content {
  padding-left: 2.0em;
}

div.exampleblock-content {
  border-left: 2px solid silver;
  padding-left: 0.5em;
}

/* IE6 sets dynamically generated links as visited. */
div#toc a:visited { color: blue; }

/* Because IE6 child selector is broken. */
div.olist2 ol {
  list-style-type: lower-alpha;
}
div.olist2 div.olist ol {
  list-style-type: decimal;
}
</style>
<title>Identifying Differentially Expressed Trinity Transcripts</title>
</head>
<body>
<div id="header">
<h1>Identifying Differentially Expressed Trinity Transcripts</h1>
</div>
<div id="preamble">
<div class="sectionbody">
<div class="para"><p>Our current system for identifying differentially expressed transcripts relies on using the EdgeR Bioconductor package. We have a protocol and scripts described below for identifying differentially expressed transcripts and clustering transcripts according to expression profiles. This process is somewhat interactive, and described are automated approaches as well as manual approaches to refining gene clusters and examining their corresonding expression patterns.</p></div>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<div class="title">Note</div>
</td>
<td class="content">We recommend generating a single Trinity assembly based on combining all reads across all samples as inputs.  Then, reads are aligned separately back to the single Trinity assembly for downstream analyses of differential expression.   If you decide to assemble each sample separately, then you'll have difficulty comparing the results across the different samples due to differences in assemlbed transcript lengths and contiguity.</td>
</tr></table>
</div>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<div class="title">Note</div>
</td>
<td class="content">If you have biological replicates, align each replicate set of reads to the Trinity contigs independently.  See the DESeq protocol below for analyzing differential expression leveraging biological replicates.</td>
</tr></table>
</div>
<div class="para"><p>First, join the <a href="align_visualize_quantify.html#RSEM">RSEM-estimated abundance values</a> for each of your samples by running:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/util/RSEM_util/merge_RSEM_frag_counts_single_table.pl  sampleA.RSEM.isoform.results sampleB.RSEM.isoform.results ... &gt; all.counts.matrix</tt></pre>
</div></div>
<div class="para"><p>Edit the column headers in the matrix file to your liking, since this is how the samples will be named in the downstream analysis steps.</p></div>
</div>
</div>
<h2 id="_identifying_differentially_expressed_transcripts">Identifying Differentially Expressed Transcripts</h2>
<div class="sectionbody">
<h3 id="_run_edger_no_biological_replicates">Run EdgeR (no biological replicates)</h3><div style="clear:left"></div>
<div class="para"><p>In case you need to install edgeR, visit the <a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR Bioconductor site</a>.</p></div>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<div class="title">Note</div>
</td>
<td class="content">A full example of the edgeR pipeline involving combining reads from multiple samples, assembling them using Trinity, separately aligning reads back to the trintiy assemblies, abundance estimation using RSEM, and differential expression analysis using edgeR is provided at: TRINITY_RNASEQ_ROOT/sample_data/test_full_edgeR_pipeline</td>
</tr></table>
</div>
<div class="para"><p>Extract the transcript length information from one of the RSEM output files like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>cat sampleA.RSEM.isoforms.results | cut -f1,3,4 &gt; Trinity.trans_lengths.txt</tt></pre>
</div></div>
<div class="para"><p>Using the all.counts.matrix file created above, perform TMM (trimmed mean of M-values) normalization and identify differentially expressed transcripts resulting from pairwise comparisons among the samples like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/Analysis/DifferentialExpression/run_EdgeR.pl --matrix all.counts.matrix --transcript_lengths Trinity.trans_lengths.txt --output edgeR_results_dir --no_eff_length</tt></pre>
</div></div>
<div class="para"><p>If you have only a single reference sample that you want the other samples to be compared to, as opposed to the all-vs-all comparisons, indicate the reference sample's column heading with: &#8212;reference ref_column_name as it exists in the all.counts.matrix file.</p></div>
<div class="para"><p></p></div>
<div class="para"><p><span class="image">
<img src="../images/diff_expr/MAplotSmearEdgeR.png" alt="example_edgeR_MA_plot" title="example_edgeR_MA_plot" />
</span></p></div>
<div class="para"><p>The TMM and length-normalized (FPKM) expression values are provided in a file: edgeR_results_dir/Trinity.fasta.normalized.FPKM, which can be examined using additional methods described below.</p></div>
<h3 id="_run_deseq_leveraging_biological_replicates">Run DESeq (leveraging biological replicates)</h3><div style="clear:left"></div>
<div class="para"><p>In case you need to install DESeq, visit the <a href="http://www.bioconductor.org/packages/2.6/bioc/html/DESeq.html">DESeq Bioconductor site</a>.</p></div>
<div class="para"><p>Create a text file <em>samples_described.txt</em> that describes the relationship between samples and biological replicates like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>conditionA   condA-rep1
conditionA   condA-rep2</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>conditionB   condB-rep1
conditionB   condB-rep2</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>conditionC   condC-rep1
conditionC   condC-rep2</tt></pre>
</div></div>
<div class="para"><p>Where condA-rep1, condA-rep2, condB-rep1, etc&#8230;, are all column names in the <em>all.counts.matrix</em> generated earlier (see top of page). Your sample names that group the replicates are user-defined here.</p></div>
<div class="para"><p>Now, run DESeq for each pair of conditions like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/Analysis/DifferentialExpression/run_DESeq_all_vs_all.pl 'samples_described.txt' 'all.counts.matrix'</tt></pre>
</div></div>
</div>
<h2 id="_analyzing_differentially_expressed_transcripts">Analyzing Differentially Expressed Transcripts</h2>
<div class="sectionbody">
<div class="para"><p>An initial step in analyzing differential expression is to extract those transcripts that are most differentially expressed (most significant P-values and fold-changes) and to cluster the transcripts according to their patterns of differential expression across the samples. To do this, you can run the following from within the edgeR output directory</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/Analysis/DifferentialExpression/analyze_diff_expr.pl --matrix transcript_read_counts.RAW.normalized.FPKM -P 1e-3 -C 2</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>If you used DESeq to call differentially expressed transcripts, then</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>first, perform the TMM normalization:</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/Analysis/DifferentialExpression/run_EdgeR.pl --matrix all.counts.matrix --transcripts Trinity.fasta --just_TMM --output TMM</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>and then run the analysis script, including the --DESeq option:</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/Analysis/DifferentialExpression/analyze_diff_expr.pl --matrix TMM/matrix.TMM_normalized.FPKM -P 1e-3 -C 2 --DESeq</tt></pre>
</div></div>
<div class="para"><p></p></div>
<div class="para"><p><span class="image">
<img src="../images/diff_expr/clustered_heatmap.png" alt="heatmap" title="heatmap" />
</span></p></div>
<div class="para"><p>The above is mostly just a visual reference. To more seriously study and define your gene clusters, you will need to interact with the data as described below. The clusters and all required data for interrogating and defining clusters is all saved with an R-session, locally with the file <em>all.RData</em>. This will be leveraged as described below.</p></div>
</div>
<h2 id="_automatically_defining_a_k_number_of_gene_clusters">Automatically defining a K-number of Gene Clusters</h2>
<div class="sectionbody">
<div class="para"><p>Run the command below to automatically split the data set into a set of $num_clusters (similar to k-means clustering).</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/Analysis/DifferentialExpression/define_clusters_by_cutting_tree.pl -K $num_clusters</tt></pre>
</div></div>
<div class="para"><p></p></div>
<div class="para"><p>To plot the mean-centered expression patterns for each cluster, visit that directory and run:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/Analysis/DifferentialExpression/plot_expression_patterns.pl subcluster_*</tt></pre>
</div></div>
<div class="para"><p>This will generate a summary image file: my_cluster_plots.pdf, as shown below:</p></div>
<div class="para"><p><span class="image">
<img src="../images/diff_expr/expression_profiles_for_clusters.png" alt="expression_profiles_for_clusters" title="expression_profiles_for_clusters" />
</span></p></div>
</div>
<h2 id="_manually_defining_gene_clusters">Manually Defining Gene Clusters</h2>
<div class="sectionbody">
<div class="para"><p>Manually defining your clusters is the best way to organize the data to your liking. This is an interactive process. Fire up R from within your output directory, being sure it contains the <em>all.RData</em> file, and enter the following commands:
R</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>load("all.RData")</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>source("TRINITY_RNASEQ_ROOT/Analysis/DifferentialExpression/R/manually_define_clusters.R")</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>manually_define_clusters(hc_genes, centered_data)</tt></pre>
</div></div>
<div class="para"><p>This should yield a display containing the hierarchically clustered genes, as shown below:</p></div>
<div class="para"><p><span class="image">
<img src="../images/diff_expr/hcl_genes_tree.png" alt="expression_hcl_tree" title="expression_hcl_tree" />
</span></p></div>
<div class="para"><p>Now, manually define your clusters from left to right (order matters here, so you can decipher the results later!) by clicking on the branch vertical branch that defines the clade of interest. After clicking on the branch, it will be drawn with a red box around the selected clade, as shown below:</p></div>
<div class="para"><p><span class="image">
<img src="../images/diff_expr/manually_selected_clusters.png" alt="manually_selected_hcl_clusters_from_tree" title="manually_selected_hcl_clusters_from_tree" />
</span></p></div>
<div class="para"><p>Right click with the mouse (or double-touch a touchpad) to exit from cluster selection.</p></div>
<div class="para"><p>The clusters as selected will be written to a subdirectory manually_defined_clusters_$count_clusters, and exist in a format similar to the automated-selection of clusters described above. Likewise, you can generate plots of the expression patterns for each cluster using the <em>plot_expression_patterns.pl</em> script.</p></div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2012-10-05 12:03:04 EDT
</div>
</div>
</body>
</html>
