<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.2.6" />
<style type="text/css">
/* Debug borders */
p, li, dt, dd, div, pre, h1, h2, h3, h4, h5, h6 {
/*
  border: 1px solid red;
*/
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

tt {
  color: navy;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  font-family: sans-serif;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}

div.sectionbody {
  font-family: serif;
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}

pre {
  padding: 0;
  margin: 0;
}

span#author {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  font-size: 1.1em;
}
span#email {
}
span#revision {
  font-family: sans-serif;
}

div#footer {
  font-family: sans-serif;
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
div#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
div#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

div#preamble,
div.tableblock, div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-right: 10%;
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.5em;
  margin-bottom: 2.5em;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}

div.listingblock {
  margin-right: 0%;
}
div.listingblock > div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock > div.content {
  padding-left: 2.0em;
}

div.attribution {
  text-align: right;
}
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 2px solid silver;
}

div.exampleblock > div.content {
  border-left: 2px solid silver;
  padding: 0.5em;
}

div.verseblock div.content {
  white-space: pre;
}

div.imageblock div.content { padding-left: 0; }
div.imageblock img { border: 1px solid silver; }
span.image img { border-style: none; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
div.olist > ol {
  list-style-type: decimal;
}
div.olist2 > ol {
  list-style-type: lower-alpha;
}

div.tableblock > table {
  border: 3px solid #527bbd;
}
thead {
  font-family: sans-serif;
  font-weight: bold;
}
tfoot {
  font-weight: bold;
}

div.hlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hlist td {
  padding-bottom: 15px;
}
td.hlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
}
td.hlist2 {
  vertical-align: top;
}

@media print {
  div#footer-badges { display: none; }
}

div#toctitle {
  color: #527bbd;
  font-family: sans-serif;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}
/* Workarounds for IE6's broken and incomplete CSS2. */

div.sidebar-content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}
div.sidebar-title, div.image-title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  margin-top: 0.0em;
  margin-bottom: 0.5em;
}

div.listingblock div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock-content {
  padding-left: 2.0em;
}

div.exampleblock-content {
  border-left: 2px solid silver;
  padding-left: 0.5em;
}

/* IE6 sets dynamically generated links as visited. */
div#toc a:visited { color: blue; }

/* Because IE6 child selector is broken. */
div.olist2 ol {
  list-style-type: lower-alpha;
}
div.olist2 div.olist ol {
  list-style-type: decimal;
}
</style>
<title>TransDecoder (Find Coding Regions Within Transcripts)</title>
</head>
<body>
<div id="header">
<h1>TransDecoder (Find Coding Regions Within Transcripts)</h1>
</div>
<div id="preamble">
<div class="sectionbody">
<div class="para"><p>TransDecoder identifies candidate coding regions within transcript sequences, such as those generated by de novo RNA-Seq transcript assembly using Trinity, or constructed based on RNA-Seq alignments to the genome using Tophat and Cufflinks.</p></div>
<div class="para"><p>TransDecoder identifies likely coding sequences based on the following criteria:</p></div>
<div class="ilist"><ul>
<li>
<p>
a minimum length open reading frame (ORF) is found in a transcript sequence
</p>
</li>
<li>
<p>
a log-likelihood score similar to what is computed by the GeneID software is &gt; 0.
</p>
</li>
<li>
<p>
the above coding score is greatest when the ORF is scored in the 1st reading frame as compared to scores in the other 5 reading frames.
</p>
</li>
<li>
<p>
if a candidate ORF is found fully encapsulated by the coordinates of another candidate ORF, the longer one is reported.  However, a single transcript can report multiple ORFs (allowing for operons, chimeras, etc).
</p>
</li>
<li>
<p>
<strong>optional</strong> the putative peptide has a match to a Pfam domain above the noise cutoff score.
</p>
</li>
</ul></div>
<div class="para"><p>The software was written and is maintained by Brian Haas at The Broad Institute, and is (or will be) included in other related software such as <a href="http://trinityrnaseq.sf.net">Trinity</a>, <a href="http://pasa.sf.net">PASA</a>, and <a href="http://evidencemodeler.sf.net">EVidenceModeler</a>.</p></div>
</div>
</div>
<h2 id="_obtaining_transdecoder">Obtaining TransDecoder</h2>
<div class="sectionbody">
<div class="para"><p>The latest release of TransDecoder can be found <a href="http://sourceforge.net/projects/transdecoder/">here</a>.</p></div>
</div>
<h2 id="_running_transdecoder">Running TransDecoder</h2>
<div class="sectionbody">
<h3 id="_predicting_coding_regions_from_a_transcript_fasta_file">Predicting coding regions from a transcript fasta file</h3><div style="clear:left"></div>
<div class="para"><p>The <em>transcripts_to_best_scoring_ORFs.pl</em> script is run on a fasta file containing the target transcript sequences.  The simplest usage is as follows:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>transcripts_to_best_scoring_ORFs.pl -t target_transcripts.fasta</tt></pre>
</div></div>
<div class="para"><p>If the transcripts are oriented according to the sense strand, then include the -S flag to examine only the top strand.  Full usage info is below.</p></div>
<div class="para"><p>The script generates several output files, which are described below, but the final set of candidate coding regions can be found as files <em>best_candidates.eclipsed_orfs_removed.*</em> where extensions include .pep, .cds, .gff3, and .bed</p></div>
<h3 id="_starting_from_a_genome_based_transcript_structure_gtf_file_eg_cufflinks">Starting from a genome-based transcript structure GTF file (eg. cufflinks)</h3><div style="clear:left"></div>
<div class="para"><p>The process here is identical to the above with the exception that we must first generate a fasta file corresponding to the transcript sequences, and in the end, we recompute a genome annotation file in GFF3 format that describes the predicted coding regions in the context of the genome.</p></div>
<div class="para"><p>First, convert the transcript structure GTF file to an alignment-GFF3 formatted file (this is done only because our processes operate on gff3 rather than the starting gtf file - nothing of great consequence).  Convert gtf to alignment-gff3 like so, using cufflinks GTF output as an example:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>util/cufflinks_gtf_to_alignment_gff3.pl transcripts.gtf &gt; transcripts.gff3</tt></pre>
</div></div>
<div class="para"><p>Next, construct the transcript fasta file using the genome and the transcripts.gff3 file like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>util/cufflinks_gtf_genome_to_cdna_fasta.pl transcripts.gtf test.genome.fasta &gt; transcripts.fasta</tt></pre>
</div></div>
<div class="para"><p>Now, run the process described above to generate your best candidate ORF predictions:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>transcripts_to_best_scoring_ORFs.pl -t transcripts.fasta</tt></pre>
</div></div>
<div class="para"><p>And finally, generate a genome-based coding region annotation file:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>cdna_alignment_orf_to_genome_orf.pl best_candidates.eclipsed_orfs_removed.gff3 transcripts.gff3 transcripts.fasta &gt; best_candidates.eclipsed_orfs_removed.genome.gff3</tt></pre>
</div></div>
</div>
<h2 id="_sample_data_and_execution">Sample data and execution</h2>
<div class="sectionbody">
<div class="para"><p>The sample_data/ directory includes a <em>runMe.sh</em> script that you can execute to demonstrate the entire process, starting from a cufflinks GTF file.</p></div>
</div>
<h2 id="_output_files_explained">Output files explained</h2>
<div class="sectionbody">
<div class="literalblock">
<div class="content">
<pre><tt>longest_orfs.pep   : all ORFs meeting the minimum length criteria, regardless of coding potential.
longest_orfs.gff3  : positions of all ORFs as found in the target transcripts
longest_orfs.cds   : the nucleotide coding sequence for all detected ORFs</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>longest_orfs.cds.top_500_longest   : the top 500 longest ORFs, used for training a Markov model for coding sequences.</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>hexamer.scores                     : log likelihood score for each k-mer  (coding/random)</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>longest_orfs.cds.scores            : the log likelihood sum scores for each ORF across each of the 6 reading frames
longest_orfs.cds.scores.selected   : the accessions of the ORFs that were selected based on the scoring criteria (described at top)
best_candidates.gff3               : the positions of the selected ORFs in transcripts</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>best_candidates.eclipsed_orfs_removed.pep : peptide sequences for the final candidate ORFs; all shorter candidates within longer ORFs were removed.
best_candidates.eclipsed_orfs_removed.cds  : nucleotide sequences for coding regions of the final candidate ORFs
best_candidates.eclipsed_orfs_removed.gff3 : positions within the target transcripts of the final selected ORFs
best_candidates.eclipsed_orfs_removed.bed  : bed-formatted file describing ORF positions, best for viewing using GenomeView or IGV.</tt></pre>
</div></div>
</div>
<h2 id="_viewing_the_orf_predictions_in_a_genome_browser">Viewing the ORF predictions in a genome browser</h2>
<div class="sectionbody">
<div class="para"><p><a href="http://genomeview.org">GenomeView</a> or <a href="http://www.broadinstitute.org/igv/">IGV</a> are  recommended for viewing the candidate ORFs in the context of the genome or the transcriptome.  Examples below show GenomeView in this context.</p></div>
<h3 id="_viewing_orfs_on_target_transcripts">Viewing ORFs on target transcripts</h3><div style="clear:left"></div>
<div class="literalblock">
<div class="content">
<pre><tt>java -jar $GENOMEVIEW/genomeview.jar transcripts.fasta best_candidates.eclipsed_orfs_removed.bed</tt></pre>
</div></div>
<div class="para"><p>If you lack a genome sequence and are working exclusively with the target transcripts, you can load the transcript fasta file and the ORF predictions (bed file) into GenomeView (see below).</p></div>
<div class="para"><p><span class="image">
<img src="images/genomeview_trans.png" alt="Transcript shown with predicted ORF" title="Transcript shown with predicted ORF" />
</span></p></div>
<h3 id="_viewing_orfs_in_the_context_of_the_transcript_structures_on_the_genome">Viewing ORFs in the context of the transcript structures on the genome</h3><div style="clear:left"></div>
<div class="literalblock">
<div class="content">
<pre><tt>java -jar $GENOMEVIEW/genomeview.jar test.genome.fasta transcripts.bed best_candidates.eclipsed_orfs_removed.genome.bed</tt></pre>
</div></div>
<div class="para"><p>The original cufflinks-based transcript structures are shown in black, and the predicted coding regions are shown in cyan.</p></div>
<div class="para"><p><span class="image">
<img src="images/genomeview_cufflinks.png" alt="Cufflinks trans in GenomeView with predicted ORFs" title="Cufflinks trans in GenomeView with predicted ORFs" />
</span></p></div>
</div>
<h2 id="_extended_usage_information_for_transcripts_to_best_scoring_orfs_pl">Extended usage information for transcripts_to_best_scoring_ORFs.pl</h2>
<div class="sectionbody">
<div class="literalblock">
<div class="content">
<pre><tt>############################# Options #####################################################################
#
# ** Required:
#
# -t &lt;string&gt;                            transcripts.fasta
#
# ** Optional:
#
# -m &lt;int&gt;                               minimum protein length (default: 100)
#
# --search_pfam &lt;string&gt;                 /path/to/pfam_db.hmm to search
#                                        using hmmscan (which should be accessible via your PATH setting)
#
# -G &lt;string&gt;                            genetic code (default: universal, options: Euplotes, Tetrahymena, Candida, Acetabularia)
#
#
# -h                                     print this option menu and quit
# -v                                     verbose
#
# -S                                     strand-specific (only analyzes top strand)
# -T &lt;int&gt;                               top longest ORFs to train Markov Model (hexamer stats) (default: 500)
#
# --CPU &lt;int&gt;                            number of threads to use (passed on to hmmscan)  (default: 2)
#
# --retain_long_orfs &lt;int&gt;               retain all ORFs found that are of minimum length in nucleotides (default: 900) so 300aa
#
##############################################################################################################</tt></pre>
</div></div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2012-08-15 21:03:10 EDT
</div>
</div>
</body>
</html>
